src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,115 @@
+            _.unset(tree, findClosestAncestorWithMultipleChildren(tree, ['_ROOT_', ...pathToContext(descendant.path)]).path)
+            descendantReplaced = true
+          }
+          else {
+            const updatedDescendantPath = newPath.concat(descendant.path.slice(newPath.length))
+            _.unset(tree, ['_ROOT_', ...pathToContext(descendant.path)])
+            _.set(tree, ['_ROOT_', ...pathToContext(updatedDescendantPath)], { leaf: true, lastUpdated: timestamp(), path: updatedDescendantPath })
+          }
+        })
+        if (descendantReplaced) onNodeChange(tree, oldPath, newPath) // called once again to remove merge inconsitencty that might occur while replacing descendants by ancestor
+        else _.unset(tree, oldContext)
+      }
+      else {
+        let isMerged = false // eslint-disable-line fp/no-let
+        leafNodes.forEach(descendant => {
+          const descendantContext = ['_ROOT_', ...pathToContext(descendant.path)]
+          if (descendantContext[1] === '__EM__') return // preventing thoughts at level 1 from merged to this (temporary fix)
+          const longContext = descendantContext.length > newContext.length ? descendantContext : newContext
+          const shortContext = newContext.length < descendantContext.length ? newContext : descendantContext
+          // siblings A.B.C and A.B.D ---> A.B.D
+          if (descendantContext.length === newContext.length && newContext.length - commonPath.length === 1) {
+            isMerged = true
+            _.unset(tree, descendantContext)
+            _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+          }
+          /*
+            restricting merge of direct cousins like A.B.C.F and A.B.D.E
+            merging distant relation A.B.C and A.B.D.E.F.G.H ---> A.B.D.E.F.G.H
+          */
+          else if (!(shortContext.length - commonPath.length > 1 || (newContext.length === descendantContext.length && newContext.length - commonPath.length === 2))) {
+            isMerged = true
+            _.unset(tree, shortContext.slice(0, commonPath.length + 1))
+            _.set(tree, longContext, { leaf: true, lastUpdated: timestamp(), path: newContext.length > descendantContext.length ? newPath : descendant.path })
+          }
+        })
+        // adding new thought to the tree
+        if (!isMerged) _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+      }
+    }
+  }
+}
+
+export const onNodeDelete = (tree, oldPath) => {
+  const oldContext = ['_ROOT_', ...pathToContext(oldPath)]
+  const { node: commonNode, path: commonPath } = findTreeDeepestSubcontext(tree, oldContext)
+  if (commonNode) {
+    if (oldContext.length > commonPath.length) return
+    if (commonPath.length !== 1) _.unset(tree, commonPath) // preventing deleting of _ROOT_
+    const parentPath = commonPath.slice(0, commonPath.length - 1)
+    const grandParentPath = commonPath.slice(0, commonPath.length - 2)
```

Comment:
Explicitly referencing grandparents is a bit of a red flag. What about parents of grandparents? This seems to suggest some recursive functionality is needed. It's possible I am not understanding what is happening here however, as I am still figuring out the code.

Reply:
Right now code is bit complex and it may take a while to figure out some logic just by analyzing the code. I will write down a document explaining `treeChange` and `treeDelete` in depth. I am sure it will give you a better understanding of my perspective. Then we can continue proper discussion about further changes required.

Reply:
Okay, thanks